### PR TITLE
add `kubectl apply diff-last-applied` subcommand

### PR DESCRIPTION
**Pull request by shiywang**
https://github.com/kubernetes/community/pull/287
`apply diff-last-applied` will invoke `git diff --no-index`, if there's no `git`, it will degenerate to use `diff` also you can use a customize diff tool by setting env : `KUBECTL_EXTERNAL_DIFF`
```console
[root@shiywang template]# ./kubectl apply diff-last-applied -f nginx.yaml 
diff --git a/tmp/kubectl-diff-yd44r.yaml b/tmp/kubectl-diff-0wox0.yaml
index 46aa188..d44cab2 100644
--- a/tmp/kubectl-diff-yd44r.yaml
+++ b/tmp/kubectl-diff-0wox0.yaml
@@ -8,7 +8,7 @@ spec:
   template:
     metadata:
       labels:
-        app: nginx123
+        app: nginx
     spec:
       containers:
       - image: gcr.io/google_containers/nginx-slim:0.8
[root@shiywang template]# 
```

@kubernetes/sig-cli-pr-reviews 
@mengqiy  @alexandercampbell

### CODE REVIEW

**Comment by mengqiy** [comment]
This supersedes #46560.

File: pkg/kubectl/cmd/apply_diff_last_applied.go
Comment (mengqiy) [inline pkg/kubectl/cmd/apply_diff_last_applied.go]: It should have flag `--diff-viewer` as described in the proposal.
It should take either `KUBECTL_EXTERNAL_DIFF` or `--diff-viewer`. If they are both specified, it should check if they match. If not, error out.

Comment (shiywang) [inline pkg/kubectl/cmd/apply_diff_last_applied.go]: I already add that flag, still need to add some tests and update help information, if there's some scenario I haven't consider or code could be simplify, please comment on it, thanks

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (mengqiy) [inline pkg/kubectl/cmd/util/cmdtools.go]: I don't like the style...

You may need to refactor related code.
I think you can try something like:
```
func Complete() {
  // env = GetEnv()
  // fvalue = GetFlagValue()
  // validate env and fvalue, and determine what diff to use
  // o.DiffTool = ...
}

func RunApplyDiffApplied() {
  // launch the cmd tool using o.DiffTool
}
```

If it doesn't make sense to you, let me know.

Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: I think that's even worse
```
func Complete() {
	env = GetEnv()
	fvalue = GetFlagValue()
	if len(env) == 0 && len(fvalue) > 0 {
		o.DiffTool = fvalue
	} else if len(env) > 0 && len(fvalue) == 0 {
		o.DiffTool = env
	} else if env != fvalue {
		return err
	} else { // env and fvalue both is ""  {
		o.DiffTool = checkGitDiffInstalled(gitDiff, normalDiff)
	}

}

func RunApplyDiffApplied() {
	// launch the cmd tool using o.DiffTool
        // we have to directly use CmdTools here instead of NewDefaultCmdTool
        // becase edit command rely on it
	diff := CmdTool{
		Name:  DiffCmd,
		Args:  o.DiffTools,
		Shell: false,   //not sure if we can set hard-coded false here
                //bash -c "command_string" is only for test ? if not we also should move defaultEnvShell logic to here...
	}
	
	diff.LaunchTempFiles(tmpFileSlice...)
}
```

Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: @mengqiy any ideas ?

**Comment by mengqiy** [comment]
I'm OK with support this subcommand without the `--diff-viewer` support. We can always add the flag later.
I can work on refactor the code to support the flag after this PR merged. I don't want to block you on this.
Please create an issue to track this and other remaining issues:
- `--diff-viewer` flag
- `_` is not printed correctly. (Any issue to track this?)

File: pkg/kubectl/cmd/apply_diff_last_applied.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/apply_diff_last_applied.go]: What is going on with this `usage` variable?

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: Do we really want to be this "smart" about automagically switching to a different diff when the user has explicitly specified `diff` in their environment variable?

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: There are more than three ways to create a string literal in bash. This can be trivially broken.

For instance:

```bash
MY_ENV=<<-EOF
this is a string with space characters in it
EOF
```

In general, I do not like using `strings` functions on shell-related literals. There are too many edge cases.
Is there a better way to do this?

**Review by alexandercampbell — CHANGES_REQUESTED** [review]
I think there were a couple open questions left on this from the older PR. I'll make another pass at this this afternoon.

Thanks!

File: pkg/kubectl/cmd/apply_test.go
Comment (shiywang) [inline pkg/kubectl/cmd/apply_test.go]: because I follow the previous coding style, seehttps://github.com/shiywang/kubernetes/blob/502bfdf94410f597a86a1471a945b1520979dac6/pkg/kubectl/cmd/create_secret_test.go#L43-L43, https://github.com/shiywang/kubernetes/blob/a82c4481f30e5877edd614f6dc7a323b5e4caf67/pkg/kubectl/cmd/create_test.go#L52-L52 and https://github.com/shiywang/kubernetes/blob/502bfdf94410f597a86a1471a945b1520979dac6/pkg/kubectl/cmd/create_namespace_test.go#L38-L38 ...

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: if user set `KUBECTL_EXTERNAL_DIFF=diff`, then it won't go this line because len(cmd) > 0.

File: pkg/kubectl/cmd/apply_diff_last_applied.go
Comment (shiywang) [inline pkg/kubectl/cmd/apply_diff_last_applied.go]: @mengqiy wrote this, my understanding is usage in `AddFilenameOptionFlags` can support different commands, we you add it to a new command, the only part you need to change is that `usage` variable, kind of decouple ?

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: I'm not sure how do you define broken, could you tell me more specific ? like list the steps which cause `kubectl` broken.

**Comment by shiywang** [comment]
@mengqiy @alexandercampbell sorry for my late response, just had a fever last week.

**Comment by alexandercampbell** [comment]
@shiywang oh no! Rest up and feel better :)
I'm working through your questions now.

File: pkg/kubectl/cmd/apply_test.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/apply_test.go]: Ha, good justification. I'm a little surprised that's our existing style.

File: pkg/kubectl/cmd/apply_diff_last_applied.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/apply_diff_last_applied.go]: I just don't understand why it's necessary to make it a separate variable if it's only used once as part of another string. @mengqiy can you help me understand? Thanks!

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: Something like this:

```bash
export KUBECTL_EXTERNAL_DIFF="git
diff
--no-index"

echo '$KUBECTL_EXTERNAL_DIFF = '$KUBECTL_EXTERNAL_DIFF

kubectl create deployment hello-world --image nginx
kubectl get deployment hello-world -o json > ./hello-world-deployment.json
kubectl apply diff-last-applied pods -f ./hello-world-deployment.json
```

Output:

```
$KUBECTL_EXTERNAL_DIFF = git diff --no-index
Error from server (AlreadyExists): deployments.extensions "hello-world" already exists
error: unable to launch the diff "git\ndiff\n--no-index"
```

Running `"bash -c $KUBECTL_EXTERNAL_DIFF"` in this script works just fine; it's only broken due to the custom string split code here in this function. Now obviously this is a bit of a contrived example. My point is that the shell is an extremely complicated system and has a million edge cases.

There's two ways to fix this:

1. Add '\n' to the list of things we check the string for and update the string.Split() code to split on '\n' as well.
2. Punt to the shell in all cases instead of trying to parse the string ourselves.

I would strongly prefer punting to the shell in all cases-- offload more work to the shell.
Does that make sense?

File: pkg/kubectl/cmd/apply_diff_last_applied.go
Comment (mengqiy) [inline pkg/kubectl/cmd/apply_diff_last_applied.go]: I wrote that abut one year ago. It may be not good now. Feel free to refactor it or just remove it if it is not helpful. :)

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: I can change to directly punt to the shell if using ENV, but in your contrived example, even change to punt to shell also doesn't work, I think in this case it's not an edge case, it's an invalid case for users.  if we want to fix this we only can trim the '\n' from args, but I don't think we should do that because it's invalid to contains '\n' in command.  cc @mengqiy wdyt ?

Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: Yes, agreed, the example is contrived and doesn't make sense to use in the real world.

> I think punt to the shell in all cases probably won't have problems

Can we just do that then? If newlines in KUBECTL_EXTERNAL_DIFF cause a problem, we can point to the shell.

Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: done

**Comment by shiywang** [comment]
@alexandercampbell @mengqiy seems  edit unit test is broken when change to punt to the shell directly, I will take a look and fix it tomorrow

File: pkg/kubectl/cmd/apply_test.go
Comment (liggitt) [inline pkg/kubectl/cmd/apply_test.go]: don't assume you can write to `/tmp`

File: pkg/kubectl/cmd/testdata/diff/test_diff.sh
Comment (liggitt) [inline pkg/kubectl/cmd/testdata/diff/test_diff.sh]: don't assume you can write to /tmp

**Comment by liggitt** [comment]
> seems edit unit test is broken when change to punt to the shell directly, I will take a look and fix it tomorrow

it looks like this change actually regressed function in several scenarios:
* editing introduces a syntax error (it should reopen, indicate the error, allow correcting, and complete successfully)
* repeatedly submitting content with an error (it should reopen and indicate the errors)
* creating a list (it doesn't look like any objects other than the first are edited/created)

when refactoring the edit commands, I would strongly suggest making small changes and verifying the unit test suite passes after each change.

**Comment by shiywang** [comment]
@liggitt @alexandercampbell  unit test fail already fixed, PTAL

**Comment by shiywang** [comment]
@liggitt @alexandercampbell  unit test fail already fixed in my last commit :https://github.com/kubernetes/kubernetes/pull/49174/commits/27af0438a2420ea402880c6763045cf63471d09f , PTAL

**Comment by shiywang** [comment]
@liggitt @mengqiy @alexandercampbell @pwittrock mind have another round of review ? I think this is almost ready.

**Comment by alexandercampbell** [comment]
@shiywang sure, reviewing

File: pkg/kubectl/cmd/apply_view_last_applied.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/apply_view_last_applied.go]: Isn't there a String() method on prettyJSON you can use?

File: pkg/kubectl/cmd/apply_view_last_applied.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/apply_view_last_applied.go]: Have you tested this? I'm surprised you don't have to do

```go
prettyJSON := new(bytes.Buffer)
```

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: What is the purpose of this? A docstring on this function would be helpful.

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: Opening which file? A little extra info in this log would be helpful.

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: You have some specific information here. Why not provide those details?

```go
return fmt.Errorf("binary %q could not be found: %v", t.Name, err)
```

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: I would reword this comment to say

```go
// The diff command returns 1 when there is a difference between the files.
```

Are you sure about this behavior is present across all the platforms we support? How can you be sure that `diff` didn't encounter an actual error (for example, permission denied)?

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: Why not do this?

```go
return fmt.Errorf("error running %v %v: %v", t.Name, cmdPrefix, err)
```

In cases where you have access to the error, it's better to include some information from that error in your result.

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: This is improved over the previous iteration. I'm still not a big fan but I won't block this from going in.

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: I think I had a comment about how this should be `unix` rather than `linux`, due to our support of MacOS.

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: I think it would be cleaner to have a function `func gitIsInstalled() bool` and do an `if` statement on that basis.

File: pkg/kubectl/cmd/util/diff/diffoptions.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/diff/diffoptions.go]: This seems to mixing CLI concerns with logical concerns about formatting the input.
In my opinion `o` should have been validated before this function is called.

**Review by alexandercampbell — COMMENTED** [review]
Made another pass. Thanks!

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: I think `t.Args` contains the filename

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: I add an error message at the end

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: I don't think `linux` is included in  `unix`.  also this var maybe some historical reason, maybe we didn't support MacOS at that time, I don't know

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: don't you think those are symmetric ?
```
	case EditorCmd:
		cmd = platformize(defaultEditor, windowsEditor)
	case DiffCmd:
		cmd = checkGitDiffInstalled(gitDiff, normalDiff)
```
I don't think we need to change here.

File: pkg/kubectl/cmd/util/diff/diffoptions.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/diff/diffoptions.go]: so do you think like this? 
```func FormatConvert(input []byte, o string) ([]byte, error) {
	switch o {
	case "json":
		var prettyJSON bytes.Buffer
		err := json.Indent(&prettyJSON, input, "", "  ")
		if err != nil {
			return nil, err
		}
		return prettyJSON.Bytes(), nil
	case "yaml":
		return yaml.JSONToYAML(input)
	}
	return nil, nil   //return nil at the end ?
}
```
also I found a bug in set-last-applied, please also take a look https://github.com/kubernetes/kubernetes/pull/50573

Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/diff/diffoptions.go]: Personally I would panic at the end of the function rather than silently returning nil.

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (alexandercampbell) [inline pkg/kubectl/cmd/util/cmdtools.go]: A better name might be `ifGitInstalled(then, else string) string`
Thoughts?

**Comment by alexandercampbell** [comment]
Thanks for addressing all those comments! Looks like merge conflicts have popped up.

File: pkg/kubectl/cmd/util/diff/diffoptions.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/diff/diffoptions.go]: I think we have a convention that do not add panic in code ? if so I think this is ok, because silently returning nil and move the Output check outside this function will make apply_set_last_applied.go looks like this
```
			if o.Output != "json" || o.Output != "yaml" {
				return fmt.Errorf("unexpected -o output mode")
			}
			out, err := diff.FormatConvert(patch.Patch, o.Output)
			if err != nil {
				return err
			}
```
which I think is just have some redundant validate code.

File: pkg/kubectl/cmd/util/cmdtools.go
Comment (shiywang) [inline pkg/kubectl/cmd/util/cmdtools.go]: ok

**Comment by alexandercampbell** [comment]
Are you running `./hack/update-bazel.sh`? Some of the BUILD changes look a little unusual.

**Comment by shiywang** [comment]
> Are you running ./hack/update-bazel.sh? Some of the BUILD changes look a little unusual.

yes, I shouldn't run that ?

**Comment by alexandercampbell** [comment]
`./hack/update-bazel.sh` is the recommended method. I was just asking because lately I was seeing some weird issues with it.
Looks like the tests are running correctly but there's a bug in the test.

**Comment by shiywang** [comment]
@mengqiy I already add back the BUILD file, do you know what the problem of this ?
```
W0831 09:36:47.961] ERROR: /workspace/k8s.io/kubernetes/pkg/kubectl/cmd/util/diff/BUILD:3:1: Target '//pkg/kubectl/resource:go_default_library' is not visible from target '//pkg/kubectl/cmd/util/diff:go_default_library'. Check the visibility declaration of the former target if you think the dependency is legitimate

W0831 09:36:48.195] ERROR: Analysis of target '//cmd/clicheck:clicheck' failed; build aborted

W0831 09:36:48.237] INFO: Elapsed time: 4.597s
W0831 09:36:48.238] FAILED: Build did NOT complete successfully (1 packages loaded)
```

**Comment by zjj2wry** [comment]
Some packages limit who can import them. You may need to explicitly add your new package to the list of packages (in the source package) that can consume it.

File: pkg/kubectl/cmd/util/BUILD
Comment (liggitt) [inline pkg/kubectl/cmd/util/BUILD]: Please don't undo this. This is working to snip dependencies to move kubectl out of the repo

**Comment by shiywang** [comment]
I'm pretty sure github has some bugs, first some commit I push didn't work, and I didn't do any squash operation... close this issue and open a new one

**Comment by shiywang** [comment]
https://status.github.com/messages/2017-10-18

**Comment by redbaron** [comment]
Why is it closed? :( Is  there another PR where work continues?

**Comment by shiywang** [comment]
@redbaron 
the last conversation is in here https://github.com/kubernetes/kubernetes/pull/54067
> @shiywang I am going to close this for now since we are in the process of rewriting apply.

as @pwittrock mentioned, and we have a new command `kubectl aplpha diff` in here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/alpha.go

**Pull request closed without merging**
